staging/kos/pkg/registry/network/subnet/strategy.go

```diff
@@ -145,135 +126,83 @@ func (*subnetStrategy) AllowUnconditionalUpdate() bool {
+		// we're here either such checks are disabled or the VNI of the subnet
+		// under validation is out of range (and the checks on conflicts make
+		// sense so long as the VNI is in range). Hence we return immediately.
+		return errs
 	}
 
-	var allSubnetsWithSameVNI []*network.Subnet
-	for _, aSubnet := range allSubnets {
-		if aSubnet.Spec.VNI == vni {
-			allSubnetsWithSameVNI = append(allSubnetsWithSameVNI, aSubnet)
-		}
-	}
-	return allSubnetsWithSameVNI, nil
-}
-
-func (s *subnetStrategy) checkSameVNISameNs(thisSubnet *network.Subnet, allSubnetsWithSameVNI []*network.Subnet) field.ErrorList {
-	errs := field.ErrorList{}
-
-	// TODO currently thisSubnet is checked against all the subnets with the
-	// same VNI. This is a sloppy attempt to deal with the fact that two subnets
-	// with same VNI but different namespaces created concurrently could both be
-	// created successfully. By checking all the subnets with the same VNI
-	// everytime a new subnet is created, we detect such anomalies. This is
-	// sloppy because if no new subnet with the given VNI is created anomalies
-	// are not detected. Also, the way the cluster is notified of the anomaly is
-	// through failure of the new subnet creation. But the anomaly is a general
-	// one concerning all the subnets with the given VNI. Creating an event
-	// could be more appropriate. When (if) we'll switch to a correct
-	// enforcement of the same VNI/same Namespace rule, checking thisSubnet
-	// against any of the already existing subnets with the same VNI, if there
-	// are any, will be enough (and more efficient).
-	for _, aSubnet := range allSubnetsWithSameVNI {
-		if aSubnet.Namespace != thisSubnet.Namespace {
-			errMsg := fmt.Sprintf("subnet %s/%s has the same VNI but different namespace with respect to subnet %s/%s. "+
-				"Subnets with the same VNI MUST reside in the same namespace",
-				thisSubnet.Namespace,
-				thisSubnet.Name,
-				aSubnet.Namespace,
-				aSubnet.Name)
-			errs = append(errs, field.Forbidden(field.NewPath("metadata", "namespace"), errMsg))
-		}
+	// Retrieve the summaries of all the other existing subnets with the same
+	// VNI so that we can check for conflicts.
+	potentialRivals, err := ss.subnetsWithVNI(fmt.Sprint(subnetSummary.VNI))
+	if err != nil {
+		return append(errs, field.InternalError(nil, err))
 	}
 
-	return errs
-}
-
-func (s *subnetStrategy) checkCIDRDoesNotOverlap(thisSubnet *network.Subnet, allSubnetsWithSameVNI []*network.Subnet) field.ErrorList {
-	errs := field.ErrorList{}
-
-	// No need to check if parse was successful, as the check has been already
-	// preformed in method checkCIDRFormat.
-	thisSubnetParsed := parseSubnet(thisSubnet)
-
-	for _, aSubnet := range allSubnetsWithSameVNI {
-		if thisSubnet.UID == aSubnet.UID {
-			// Do not compare the subnet under validation against itself (needed
-			// in case of updates).
+	// Check whether there are Namespace and CIDR conflicts with other subnets.
+	for _, pr := range potentialRivals {
```

Comment:
Why do we have two for loops, one making a list and another analyzing each member?  It would be simpler to have just one for loop.